wagtail/wagtail
# Remove xregexp (polyfill) & replace urlify (Django port) vendor library

With #11161 merged & #11157 coming (modeladmin removal), we can now start to deprecate the global `window.URLify` and remove the legacy Regex polyfill (62KB of JS no longer needed).


## Overview

- Move vendor URLify to a TypeScript port of Django's URLify function.
- Add legacy support for window.URLify with allowUnicode as a param.
- Add release notes details for both `window.URLify` (deprecated) & `window.XRegExp` (removed).
- Update JSDoc to add more context to the port and also to the deprecated function.
- Gets us a step closer to finishing #2936 (Stop importing _editor_js.html everywhere).

## Implementation details

## About URLify `allow_unicode`

The Slug widget usage of the urlify does **not** use the Django port URLify `allow_unicode`, but it turns out this is a bug. See https://github.com/wagtail/wagtail/pull/11168#discussion_r1542436595

I have **NOT fixed this bug** in this PR but instead will raise a new bug and a new PR, as we may want to backport the fix to previous stable branches.

https://github.com/wagtail/wagtail/issues/11828


### About 'stopwords'

Our comments for `urlify` indicated that this function removes stopwords, this is not true and as far as I can tell has not been the case for many years. It only cleans the string and replaces unicode like characters with their mapped equivalent (e.g. `'À' -> 'A'`).

See https://code.djangoproject.com/ticket/11157 & https://github.com/wagtail/wagtail/pull/6410 & https://github.com/wagtail/wagtail/pull/4884

As such, I have removed this reference in the comments.

## Testing

- Check that `window.URLify` is a valid global function & accepts three args.
- Check that the page edit (new and existing pages) correctly changes the slug field when it's interacted with or changed when in sync and NOT published.
- See detailed testing for slugs on https://github.com/wagtail/wagtail/pull/10568 if needed.

<img width="1730" alt="Screenshot 2023-11-01 at 8 40 11 pm" src="https://github.com/wagtail/wagtail/assets/1396140/da0b6737-8aa3-4990-b02f-851c996e0f5b">


## Checks
